Gems/Atom/RHI/DX12/Code/Source/RHI/PipelineLibrary.cpp

```diff
@@ -200,23 +205,39 @@ namespace AZ
 
         RHI::ResultCode PipelineLibrary::MergeIntoInternal([[maybe_unused]] AZStd::array_view<const RHI::PipelineLibrary*> pipelineLibraries)
         {
+#if defined(USE_PIX) || defined(USE_RENDERDOC)
+            // StorePipeline api does not function properly if Pix or RenderDoc is enabled
+            return RHI::ResultCode::Fail;
+#else
+
 #if defined (AZ_DX12_USE_PIPELINE_LIBRARY)
             AZStd::lock_guard<AZStd::mutex> lock(m_mutex);
+            size_t currentSize = m_pipelineStates.size();
 
             for (const RHI::PipelineLibrary* libraryBase : pipelineLibraries)
             {
                 const PipelineLibrary* library = static_cast<const PipelineLibrary*>(libraryBase);
 
                 for (const auto& pipelineStateEntry : library->m_pipelineStates)
                 {
-                    m_library->StorePipeline(pipelineStateEntry.first.c_str(), pipelineStateEntry.second.get());
+                    if (m_pipelineStates.find(pipelineStateEntry.first) == m_pipelineStates.end())
+                    {
+                        m_library->StorePipeline(pipelineStateEntry.first.c_str(), pipelineStateEntry.second.get());
 
-                    m_pipelineStates.emplace(pipelineStateEntry.first, pipelineStateEntry.second);
+                        m_pipelineStates.emplace(pipelineStateEntry.first, pipelineStateEntry.second);
+                    }
                 }
             }
-#endif
 
-            return RHI::ResultCode::Success;
+            if (m_pipelineStates.size() != currentSize)
+            {
+                return RHI::ResultCode::Success;
+            }
```

Comment:
I saw the function comment and this does reflect the behavior that: return failure when as if the function is never called.
But I see the actual usage will be affected if all merged pipeline states are duplicated, resulting in returning a nullptr. Like in "PipelineStateCache::GetLibrarySerializedData". Do you think we should change the comment instead?

Reply:
I cleaned up the abstraction to address this concern.